### PR TITLE
Fix parsing minified code with PEG (#363)

### LINKED ISSUE #363: JS Minification during React Build throws off eel.init
I'm building an app using React for the Front End after the example of @KyleKing 
During development everything works perfectly but upon running `npm run build` and even `npm run build -- --profile` the script is not able to parse the enclosed JS functions.

Here the code in App.js:
```
function show_log(msg) {
    window.AppComponent.showLog(msg)
}

window.eel.expose(show_log, 'show_log')
```

Here's the minified version after running `npm run build -- --profile`
```
we.set_host("ws://localhost:8080"), window.eel.expose((function show_log(e) {
            window.AppComponent.showLog(e)
        }), "show_log");
```

I don't know what else to do on the js side but I think from the Python side there are two possible solutions:

1. Adapt `eel.init` to accept an array of function names to accept as default functions to be registered (and replace the assert error with a logging statement)
2. Adapt the regex for finding the js functions → a possible alternative pattern would be: `eel\.expose\(.*[;\)]` to then pass into a more exhaustive isolation script 
I guess this would be an issue where @samuelhwilliams would need to have a say in.

Please let me know if there are specific things I can do on the JS side to avoid having to adjust the package.

### PR DESCRIPTION
**Pull request by KyleKing**
Resolves #363 

Particularly in the React example, but anywhere where JS is minified, you can end up with a variety of static JS code to parse. Currently the regex will fail because it only grabs the string between `eel.expose(` and the first `)`

> `eel.expose(func_name)`
> `window.eel.expose(w,"say_hello_js")`
> `window.eel.expose(function(e){console.log(e)},"show_log_alt")` (AssertionError: `function(`)

Alternatively, this PR demonstrates how pyparsing could be used to parse all of the above cases suppressing the nested JavaScript function if found (`pp.NestedExpr()`)

I haven't thoroughly tested these changes and only ran examples 1, 4, and 7 to quickly verify that it works for most cases. I would like to at least test with Electron and think more about what kind of edge cases could happen

### CODE REVIEW

**Comment by KyleKing** [comment]
@samuelhwilliams, I added tests and think this PR is ready for review

File: tests/unit/test_eel.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    test_cases = [
+        ('eel.expose(w,"say_hello_js")', ['"say_hello_js"']),
+        ('eel.expose(function(e){console.log(e)},"show_log_alt")', ['"show_log_alt"']),
+        (' \t\nwindow.eel.expose((function show_log(e) {console.log(e)}), "show_log")\n', ['"show_log"']),
+        ((INIT_DIR / 'minified.js').read_text(), ['"say_hello_js"', '"show_log_alt"', '"show_log"']),
+        ((INIT_DIR / 'sample.html').read_text(), ['say_hello_js']),
+    ]
+    for (js_code, expected_matches) in test_cases:
+        matches = eel.EXPOSED_JS_FUNCTIONS.parseString(js_code).asList()
+        assert matches == expected_matches, f'Expected {expected_matches} (found: {matches}) in: {js_code}'
```
Comment (samuelhwilliams) [inline tests/unit/test_eel.py:21]: Could we please use pytest parameterization here for the test cases? :)

**Comment by samuelhwilliams** [comment]
Hey @KyleKing - this looks like a great upgrade to Eel's parsing stuff in principle. I'm going to drop some comments on it over the next day or two as/when I can - feel free not to start responding until I say I've finished so that it's not too bitty for you.

Sorry it's taken so long.

**Comment by KyleKing** [comment]
No problem, I was up this morning and that was an easy fix. Let me know if there are other improvements!

File: eel/__init__.py [hunk 2]
```diff
@@ -83,6 +84,21 @@ def decorator(function):
         return function
 
 
+# PyParsing Grammer for parsing exposed functions in JavaScript code
```
Comment (samuelhwilliams) [inline eel/__init__.py:87]:
```suggestion
# PyParsing grammar for parsing exposed functions in JavaScript code
```

File: tests/unit/test_eel.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+from pathlib import Path
```
Comment (samuelhwilliams) [inline tests/unit/test_eel.py:1]: This looks like an unused import - could it be removed please?

I need to set up some linting checks...

File: tests/unit/test_eel.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+import eel
+import pytest
+from tests.utils import TEST_DATA_DIR
+
+INIT_DIR = TEST_DATA_DIR / 'init_test'
+"""Directory for testing eel.__init__."""
```
Comment (samuelhwilliams) [inline tests/unit/test_eel.py:8]:
```suggestion
# Directory for testing eel.__init__.
INIT_DIR = TEST_DATA_DIR / 'init_test'
```

Just a minor nit to have the comment before the related code and use `#` for non-docstring comments.

File: tests/utils.py [hunk 5]
```diff
@@ -3,9 +3,13 @@
 import subprocess
 import tempfile
 import time
+from pathlib import Path
 
 import psutil
 
+TEST_DATA_DIR = Path(__file__).parent / 'data'
+"""Path to the test data folder."""
```
Comment (samuelhwilliams) [inline tests/utils.py:11]:
```suggestion
# Path to the test data folder.
TEST_DATA_DIR = Path(__file__).parent / 'data'
```

Just the same here as well :)

File: tests/unit/test_eel.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+
+@pytest.mark.parametrize('js_code, expected_matches', [
+    ('eel.expose(w,"say_hello_js")', ['"say_hello_js"']),
+    ('eel.expose(function(e){console.log(e)},"show_log_alt")', ['"show_log_alt"']),
+    (' \t\nwindow.eel.expose((function show_log(e) {console.log(e)}), "show_log")\n', ['"show_log"']),
+    ((INIT_DIR / 'minified.js').read_text(), ['"say_hello_js"', '"show_log_alt"', '"show_log"']),
+    ((INIT_DIR / 'sample.html').read_text(), ['say_hello_js']),
```
Comment (samuelhwilliams) [inline tests/unit/test_eel.py:16]: Could we possibly extend this to also check the remaining two files in that dir - App.tsx and hello.html?

**Comment by samuelhwilliams** [comment]
Hey @KyleKing - I'm done reviewing now. There's a few minor comments above but it looks really good on the whole. Once you've finished updating things, could you rebase and sanitise the commits a little (i.e. probably just one commit makes sense here) - and then I'll merge.

Thanks so much for this!

File: eel/__init__.py (hunk 2 above)
Comment (KyleKing) [reply eel/__init__.py:87]: 🤦

**Comment by KyleKing** [comment]
@samuelhwilliams I made all of the changes. Thanks for the edits

I'm not sure how to squash/rebase the commits without force pushing to Github or creating a new branch and starting a new PR. Do you have the option to squash the commits below? https://github.blog/2016-04-01-squash-your-commits/

**Comment by samuelhwilliams** [comment]
I can squash them, that's fine, but force pushing to your branch would also presumably have been fine. Thanks for the updates!

File: tests/unit/test_eel.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    ((INIT_DIR / 'App.tsx').read_text(), ["'say_hello_js'", "'show_log'"]),
+    ((INIT_DIR / 'hello.html').read_text(), ['say_hello_js', 'js_random']),
```
Comment (samuelhwilliams) [inline tests/unit/test_eel.py:16]: Why are some of these quoted twice and others not? Are quotes sometimes part of the captured function name? That doesn't feel right 🤔

Comment (samuelhwilliams) [reply tests/unit/test_eel.py:16]: Ah right - I see that these later get stripped of quotes. Is it possible to update the pyparsing grammar to exclude those for us, or is that too much of a pain? It would be nice to have that grammar only emit the names without us needing to do any further post-processing (but we can probably tolerate it as-is if the extra work is significant - or I can investigate it myself :))

Comment (KyleKing) [reply tests/unit/test_eel.py:16]: Yeah I hadn't look at this code in a little while so I had the same thought that the extra quotes were weird. I initially looked into handling the quotes in PEG and experimenting with quoted string, but that ended up not working as expected. I found that stripping the quotes after was just a bit cleaner

I have to wrap something up, but I can take a quick look tonight in the next hour or so. I think it *may* be straightforward to suppress the quotes as optional characters add quotes to the `excluded chars` for `pp.Word` match

Comment (KyleKing) [reply tests/unit/test_eel.py]: Got it!

```py
EXPOSED_JS_FUNCTIONS = pp.ZeroOrMore(
    pp.Suppress(
        pp.SkipTo(pp.Literal('eel.expose('))
        + pp.Literal('eel.expose(')
        + pp.Optional(
            pp.Or([pp.nestedExpr(), pp.Word(pp.printables, excludeChars=',')]) + pp.Literal(',')
        )
    )
    + pp.Suppress(pp.Regex(r'["\']?'))
    + pp.Word(pp.printables, excludeChars='"\')')
    + pp.Suppress(pp.Regex(r'["\']?\s*\)')),
)
```

Comment (samuelhwilliams) [reply tests/unit/test_eel.py]: Nice!

**Comment by ChrisKnott** [comment]
This is a really great change @KyleKing

**Comment by KyleKing** [comment]
Thanks!
